### PR TITLE
ARROW-9400: [Python] Do not depend on conda-forge static libraries in Windows wheel builds

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@kou @kszucs PTAL. I would like to merge this to get the CI fixed and then rebase https://github.com/apache/arrow/pull/8878  on top of that. There I might remove the conda dependency completely.

**Comment by xhochy** [comment]
@github-actions crossbow submit -g wheel

**Comment by xhochy** [comment]
@github-actions autotune everything

File: dev/tasks/python-wheels/win-build.bat [hunk 1]
```diff
@@ -46,32 +47,30 @@ pushd arrow\cpp\build
...
-      -DARROW_WITH_BROTLI=ON ^
       -DARROW_DATASET=ON ^
+      -DARROW_DEPENDENCY_SOURCE=BUNDLED ^
       -DARROW_FLIGHT=ON ^
       -DARROW_GANDIVA=OFF ^
-      -DARROW_LZ4_USE_SHARED=OFF ^
       -DARROW_MIMALLOC=ON ^
+      -DARROW_OPENSSL_USE_SHARED=OFF ^
       -DARROW_PARQUET=ON ^
+      -DARROW_PROTOBUF_USE_SHARED=OFF ^
```
Comment (kszucs) [inline dev/tasks/python-wheels/win-build.bat:60]: Should `ARROW_DEPENDENCY_USE_SHARED` be enough to set?

**Comment by xhochy** [comment]
@github-actions autotune

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
@kou Do you have an idea what would be causing the MinGW failures?

**Comment by pitrou** [comment]
@xhochy This looks like a protobuf version mismatch perhaps?

**Comment by xhochy** [comment]
Seems like we are building our own gRPC but then use the headers from the MinGW-provided ones, just another case of the RE2 vs re2 weirdness.

**Comment by xhochy** [comment]
> Seems like we are building our own gRPC but then use the headers from the MinGW-provided ones, just another case of the RE2 vs re2 weirdness.

Doesn't make a difference with the system gRPC 🤔

**Comment by kou** [comment]
#8908 solves the MinGW build failures. It's caused by signature change in gRPC (`grpc::protobuf::Message` -> `grpc::protobuf::MessageLite`): https://github.com/apache/arrow/pull/8908/files#diff-00ed5e81d6488cc38237bbba649ca625d679cf6035a90f15341db7d25e73602a

**Comment by xhochy** [comment]
@kou Nice, I'll then just wait on the merge of #8908  and will then rebase (and revert some of the changes here)

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 2]
```diff
@@ -2411,7 +2432,11 @@ macro(build_grpc)
   set(
     GRPC_STATIC_LIBRARY_UPB
     "${GRPC_PREFIX}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}upb${CMAKE_STATIC_LIBRARY_SUFFIX}")
-  set(GRPC_CPP_PLUGIN "${GRPC_PREFIX}/bin/grpc_cpp_plugin")
+  if(MSVC)
+    set(GRPC_CPP_PLUGIN "${GRPC_PREFIX}/bin/grpc_cpp_plugin.exe")
+  else()
+    set(GRPC_CPP_PLUGIN "${GRPC_PREFIX}/bin/grpc_cpp_plugin")
+  endif()
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:2439]: Can we use `CMAKE_EXECUTABLE_SUFFIX` here?
https://cmake.org/cmake/help/latest/variable/CMAKE_EXECUTABLE_SUFFIX.html

```suggestion
  set(GRPC_CPP_PLUGIN "${GRPC_PREFIX}/bin/grpc_cpp_plugin${CMAKE_EXECUTABLE_SUFFIX}")
```

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 3]
```diff
@@ -2465,20 +2490,35 @@ macro(build_grpc)
   string(REPLACE ";" "|" GRPC_PREFIX_PATH_ALT_SEP "${GRPC_CMAKE_PREFIX}")
 
   set(GRPC_CMAKE_ARGS
+      -DBUILD_SHARED_LIBS=OFF
       -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+      -DCMAKE_CXX_FLAGS=${GRPC_CMAKE_CXX_FLAGS}
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:2495]: Can we use `EP_COMMON_CMAKE_ARGS` for `GRPC_CMAKE_ARGS`?

It seems that https://github.com/grpc/grpc/issues/13856 has been fixed in gRPC 1.33.1 or later.

**Review by kou — APPROVED** [review]
+1

We'll merge #8908 -> #8881 -> ...

**Comment by nealrichardson** [comment]
@xhochy what's the status of this? Fixing windows wheels is one of our (several) release blockers. Is rebase all that is needed to be able to merge this?

**Comment by xhochy** [comment]
I can rebase and see whether there is progress but it might still have issues and I had hopes for #9096 to land in time.

**Comment by xhochy** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by kszucs** [comment]
INFRA needs to set a new secret for the bot https://issues.apache.org/jira/browse/INFRA-21280 

Until that please use crossbow locally: 

```bash
python dev/tasks/crossbow.py submit wheel-win-cp38
```

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by xhochy** [comment]
Something went wrong with the crossbow task. It should have triggered a build on Github Actions but it didn't create the appropriate CI files in the branch. @kszucs can you take a look?

**Comment by kszucs** [comment]
It may need more scopes, looking.

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by nealrichardson** [comment]
@kszucs can this be merged now?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 4]
```diff
@@ -2124,9 +2124,20 @@ macro(build_re2)
   set_target_properties(re2::re2
                         PROPERTIES IMPORTED_LOCATION "${RE2_STATIC_LIB}"
                                    INTERFACE_INCLUDE_DIRECTORIES "${RE2_PREFIX}/include")
+  if(NOT TARGET re2::re2)
+    add_library(re2::re2 STATIC IMPORTED)
+    set_target_properties(
+      re2::re2
+      PROPERTIES IMPORTED_LOCATION "${RE2_STATIC_LIB}" INTERFACE_INCLUDE_DIRECTORIES
+                 "${RE2_PREFIX}/include")
+  endif()
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:2133]: Can we remove this because we always add `re2::re2` library at the above line.

**Comment by kou** [comment]
@github-actions crossbow submit wheel-win-cp38

**Comment by kou** [comment]
+1

**Pull request closed without merging**